test-workloads/src/bin/guestvm.rs

```diff
@@ -319,6 +320,41 @@ fn test_memory_sharing() {
     }
 }
 
+fn test_interrupts() {
+    const INTERRUPT_ID: u64 = 3;
+
+    // Make sure we return from WFI.
+    //
+    // Safety: WFI behavior is well-defined.
+    unsafe { asm!("wfi", options(nomem, nostack)) };
+
+    // Enable external interrupt signalling so that we can check for them in SIP (and that the
+    // host sees them in HGEIP), but leave SSTATUS.SIE unset so that we don't trap.
+    //
+    // TODO: Set up and interrupt handler and enable interrupts so that we can receive injected
+    // interrupts.
+    CSR.sie.read_and_set_field(sie::sext);
+
+    // EIDELIVERY = 1
+    CSR.siselect.set(0x70);
+    CSR.sireg.set(1);
+    // EITHRESHOLD = 0
+    CSR.siselect.set(0x72);
+    CSR.sireg.set(0);
+    // EIE0[3] = 1
+    CSR.siselect.set(0xc0 + (INTERRUPT_ID / 64));
+    CSR.sireg.read_and_set_bits(1 << (INTERRUPT_ID % 64));
+
+    tee_guest::allow_external_interrupt(3).expect("GuestVm - AllowExternalInterrupt failed");
+
+    // VSIP implementation is buggy in QEMU; use VSTOPI to check that we got the interrupt instead.
```

Comment:
How does it manifest ? Have you sent a fix ?

Reply:
Mailed a fix last week: https://lore.kernel.org/qemu-devel/20221215224541.1423431-1-abrestic@rivosinc.com/

Basically the bit-masking in VSIP (and to some extent VSIE) is kinda bogus and SSIP ends up being the only readable bit.

Reply:
Thanks!